pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,601 @@
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::force_inflation_recalculation())]
+        pub fn force_inflation_recalculation(origin: OriginFor<T>) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let (max_emission, config) =
+                Self::recalculate_inflation(frame_system::Pallet::<T>::block_number());
+
+            ActiveInflationConfig::<T>::put(config.clone());
+
+            SafetyInflationTracker::<T>::mutate(|tracker| {
+                tracker.cap.saturating_accrue(max_emission);
+            });
+
+            Self::deposit_event(Event::<T>::ForcedInflationRecalculation { config });
+
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Used to check if inflation recalculation is supposed to happen on the next block.
+        fn is_recalculation_in_next_block(
+            now: BlockNumber,
+            config: &InflationConfiguration,
+        ) -> bool {
+            config.recalculation_block.saturating_sub(now) <= 1
+        }
+
+        /// Payout block rewards to the beneficiaries.
+        ///
+        /// Return the total amount issued.
+        fn payout_block_rewards() -> Balance {
+            let config = ActiveInflationConfig::<T>::get();
+
+            let collator_amount = T::Currency::issue(config.collator_reward_per_block);
+            let treasury_amount = T::Currency::issue(config.treasury_reward_per_block);
+
+            T::PayoutPerBlock::collators(collator_amount);
+            T::PayoutPerBlock::treasury(treasury_amount);
+
+            config.collator_reward_per_block + config.treasury_reward_per_block
+        }
+
+        /// Recalculates the inflation based on the total issuance & inflation parameters.
+        ///
+        /// Returns the maximum total emission for the cycle, and the new inflation configuration.
+        pub(crate) fn recalculate_inflation(now: BlockNumber) -> (Balance, InflationConfiguration) {
+            let params = InflationParams::<T>::get();
+            let total_issuance = T::Currency::total_issuance();
+
+            // 1. Calculate maximum emission over the period before the next recalculation.
+            let max_emission = params.max_inflation_rate * total_issuance;
+
+            // 2. Calculate distribution of max emission between different purposes.
+            let treasury_emission = params.treasury_part * max_emission;
+            let collators_emission = params.collators_part * max_emission;
+            let dapps_emission = params.dapps_part * max_emission;
+            let base_stakers_emission = params.base_stakers_part * max_emission;
+            let adjustable_stakers_emission = params.adjustable_stakers_part * max_emission;
+            let bonus_emission = params.bonus_part * max_emission;
+
+            // 3. Calculate concrete rewards per block, era or period
+
+            // 3.0 Convert all 'per cycle' values to the correct type (Balance).
```

Comment:
Not sure I get the comment, sorry.

I need to cast it since I'm dividing a variable with `Balance` type. Otherwise it would be `u128 / u32`.

Perquintills are only used above when calculation emission - after that point it's all `Balance` types.